lib/ansible/modules/crypto/openssl_cert.py

```diff
@@ -0,0 +1,996 @@
+  openssl_cert:
+    dest: /etc/ssl/crt/example.com.crt
+    provider: assertonly
+    keyUsage_contains:
+      - key_agreement: False
+
+- name: Ensure that the existing certificate is only used for digital signatures and encrypting other keys
+  openssl_cert:
+    dest: /etc/ssl/crt/example.com.crt
+    provider: assertonly
+    keyUsage_only:
+      - digital_signature: True
+      - key_encipherment: True
+
+- name: Ensure that the existing certificate can be used for client authentication
+  openssl_cert:
+    dest: /etc/ssl/crt/example.com.crt
+    provider: assertonly
+    extendedKeyUsage_contains:
+      - clientAuth
+
+- name: Ensure that the existing certificate can only be used for client authentication and time stamping
+  openssl_cert:
+    dest: /etc/ssl/crt/example.com.crt
+    provider: assertonly
+    extendedKeyUsage_only:
+      - clientAuth
+      - 1.3.6.1.5.5.7.3.8
+
+- name: Ensure that the existing certificate has a certain domain in its subjectAltName
+  openssl_cert:
+    dest: /etc/ssl/crt/example.com.crt
+    provider: assertonly
+    subjectAltName_contains:
+      - www.example.com
+      - test.example.com
+'''
+
+
+RETURN = '''
+cert_path:
+    description: Path to the generated Certificate
+    returned: changed or success
+    type: string
+    sample: /etc/ssl/crt/www.ansible.com.crt
+'''
+
+
+import datetime
+from random import randint
+import re
+import os
+
+try:
+    from OpenSSL import crypto
+    from OpenSSL import __version__ as opensslversion
+except ImportError:
+    pyopenssl_found = False
+else:
+    pyopenssl_found = True
+
+try:
+    from cryptography import x509
+    from cryptography.hazmat.backends import default_backend
+    from OpenSSL import __version__ as cryptographyversion
+except ImportError:
+    cryptography_found = False
+else:
+    cryptography_found = True
+
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.pycompat24 import get_exception
+
+
+class CertificateError(Exception):
+    pass
+
+
+class Certificate(object):
+
+    def __init__(self, module):
+        self.state = module.params['state']
+        self.dest = module.params['dest']
+        self.force = module.params['force']
+        self.provider = module.params['provider']
+        self.signature_algorithms = module.params['signature_algorithms']
+        self.privatekey_path = module.params['privatekey']
+        self.issuer = module.params['issuer']
+        self.subject = module.params['subject']
+        self.has_expired = module.params['has_expired']
+        self.version = module.params['version']
+        self.csr_path = module.params['csr']
+        self.module = module
+        self.changed = True
+        self.warnings = []
+
+    def generate(self):
+        '''Generate the certificate signing request.'''
+        # Implemented in child classes
+        pass
+
+    def remove(self):
+        '''Remove the certificate'''
+
+        try:
```

Comment:
One specific case is missing in the  try/catch. See https://github.com/ansible/ansible/commit/ed15a0e9a7bf237786634cc5dd89d9448031c86f